### PR TITLE
Disallow adding to score after the player has failed

### PR DESCRIPTION
**Pull request by frenzibyte**
Mostly for the health bar to not increment, but thought it's actually weird for the score to increment even though the player is in a failed state, as for real-time players and reviving logic, can be done by adding a boolean property (`Revivable`) that allows that or something.

### CODE REVIEW

**Review by peppy — DISMISSED** [review]
i think this seems ok

**Comment by smoogipoo** [comment]
This should have an inverse operation inside `RevertResult()`. A simple way of doing this is to store another property inside the result `HasFailedAtJudgement`, and use that to decide whether the judgement should be reverted.

File: osu.Game/Rulesets/Scoring/ScoreProcessor.cs [hunk 1]
```diff
@@ -316,6 +316,9 @@ private void revertResult(JudgementResult result)
         /// <param name="result">The <see cref="JudgementResult"/> to apply.</param>
         protected virtual void ApplyResult(JudgementResult result)
         {
+            if (HasFailed)
+                return;
+
             result.ComboAtJudgement = Combo.Value;
             result.HighestComboAtJudgement = HighestCombo.Value;
             result.HealthAtJudgement = Health.Value;
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Scoring/ScoreProcessor.cs]: Along with the previous comment, the return should be _after_ all these are set.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by iiSaLMaN** [comment]
> This should have an inverse operation inside `RevertResult()`. A simple way of doing this is to store another property inside the result `HasFailedAtJudgement`, and use that to decide whether the judgement should be reverted.

I don't quite understand you, you mean to invoke `RevertResult` if the player has failed after `NewResult` is finished? or invoke it somewhere else?

**Comment by smoogipoo** [comment]
No, RevertResult is used to un-apply the results while rewinding. If ApplyResult has been invoked, RevertResult will also be invoked for the same result during rewind.

We may at some point, implement some mechanic that rewinds to a non-failed state.

**Comment by smoogipoo** [comment]
@iiSaLMaN I've refactored it as I originally envisioned it: https://github.com/ppy/osu/pull/5637/commits/88fa06efbadc23501209021ad4df1436bd4acb7f

Note that I've removed the `HasFailed` revert since it isn't handled well, but I've added a todo for it.

@peppy requesting review

**Review by peppy — DISMISSED** [review]

**Pull request merged**